### ISSUE TITLE
Usage question

### ISSUE THREAD

**Issue by ArnoLP** [AFaust/simple-content-stores#2]
Hello Axel,

This is not an issue, but a question about how to use this add-on.

At first, to be sure I understood it correctly, please let me rephrase what I understood :

This add-on can define custom content stores, these custom content stores getting potentially specific behaviours (compression as an example)
As an example, if I want that a specific site document library is stored in another content store than the default Alfresco one, this addon can define this custom content store.

Am I right ?

If OK, then I don't see with the provided example and information how I should configure the add-on. Sorry for this.

Keeping my example (specific site document library should be stored in another content store), what should be the configuration items ?

In fact, I would need to define (at least) the selector property, the root directory of the custom content store, and the value of the selector property that would indicate which rootDirectory it should use.

If I am right, how should I configure this ?

I found the related information in 2 different store types : the "selectorPropertyStore" and the "standardFileStore" ones, and so I don't see how to link them.

Or maybe I really didn't understand how it behaves and how it should be configured. Sorry if it's the case.

Can you please help ?

Thanks in advance.

Kind Regards,

Arnaud.

**Comment by AFaust** [comment]
You are right, the addon provides the ability to configure / define custom content stores. Your example of a separate store per site (or even per site container) could be achieved with the selectorPropertyStore, provided each document is assigned a property that reflects the site / site container, e.g. each document has a property "my:site". This property would then be configured as the "selectorPropertyName" and the names of the sites would be keys in the "map.storeBySelectorPropertyValue" configuration.

The following is a theoretical configuration (not tested / verified; but based on the documented options):
```
simpleContentStores.enabled=true
simpleContentStores.rootStore=siteRoutingStore
simpleContentStores.customStores=siteRoutingStore,siteAB1Store,siteAB2Store,siteAB3Store

simpleContentStores.customStore.siteRoutingStore.type=selectorPropertyStore
simpleContentStores.customStore.siteRoutingStore.value.selectorClassName=my:siteContent
simpleContentStores.customStore.siteRoutingStore.value.selectorPropertyName=my:siteShortName
# fallback to standard content store
simpleContentStores.customStore.siteRoutingStore.ref.fallbackStore=fileContentStore
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB1=siteAB1Store
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB2=siteAB2Store
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB3=siteAB3Store
# move stores whenthe value of my:siteShortName changes
simpleContentStores.customStore.siteRoutingStore.value.moveStoresOnChange=true

simpleContentStores.customStore.siteAB1Store.type=standardFileStore
simpleContentStores.customStore.siteAB1Store.value.protocol=site-ab1-store
simpleContentStores.customStore.siteAB1Store.value.rootDirectory=${dir.root}/site-ab1-contentstore

simpleContentStores.customStore.siteAB2Store.type=standardFileStore
simpleContentStores.customStore.siteAB2Store.value.protocol=site-ab2-store
simpleContentStores.customStore.siteAB2Store.value.rootDirectory=${dir.root}/site-ab2-contentstore

simpleContentStores.customStore.siteAB3Store.type=standardFileStore
simpleContentStores.customStore.siteAB3Store.value.protocol=site-ab3-store
simpleContentStores.customStore.siteAB3Store.value.rootDirectory=${dir.root}/site-ab3-contentstore
```

This setup requires that you have a model defining the aspect + property used to select the store, and a policy / rule that actually assignes the name of the site as the value of the property on creation / move of a document.
The drawback of this approach is that you would need to shut-down and restart evertime you want to create a new site with its own, separate content store.

This is definitely a use case that might be relevant for other users as well. I will think about a potential new content store that could serve this without requiring a custom model, policy or reconfiguring/restarting the server.

**Comment by ArnoLP** [comment]
Hello,

I tried with the exact example you provided, and it doesn't behave as expected.

When I assign the name of the site as the value of the property, then it doesn't change anything. The content is still stored in the default contentstore, and not the specific one.

Could you help me, and provide me with some hints which would help me to analyze where the issue is ?

Thanks in advance.

**Comment by AFaust** [comment]
Hello Arno,

it would be great if you could somehow provide the complete model + behaviour + content store setup so I could take a closer look at potential problems. My guess at the moment is that there is something with the model + behaviour that may end up interfering with correct behaviour.

Regardless of the problem / error you may be having, I have started work on a more convenient content store implementation and already have a working albeit not yet 100% complete state pushed to the GitHub repository. I have sucessfully used the following configuration to store the contents of a new site "my-store" in a separate directory:

```
simpleContentStores.enabled=true

simpleContentStores.rootStore=siteFileContentStore
simpleContentStores.customStores=siteFileContentStore

simpleContentStores.customStore.siteFileContentStore.type=siteAwareMultiDirectoryFileStore
# default root directory for all other content
simpleContentStores.customStore.siteFileContentStore.value.rootDirectory=${dir.contentstore}

# site specific root directory + protocol
simpleContentStores.customStore.siteFileContentStore.map.rootAbsolutePathsBySite.value.my-site=${dir.root}/my-site-contentstore
simpleContentStores.customStore.siteFileContentStore.map.protocolsBySite.value.my-site=my-site-store
```

This new store removes the need to define a custom model and code a behaviour to map the site name to a property. I still have to adapt some of the other content stores to be fully compatible with a new "context" concept I had to introduce to make this work, and have yet to implement the move/copy of content between directories if a node is moved / copied in/out of or between sites.
The new "context" concept will also allow me to build a couple of other simple multi-directory / routing content stores and may also lead to some support of multi-tenancy.

**Comment by ArnoLP** [comment]
Hello Axel,

Thanks for your feedback.

here is my configuration :

_**model :**_
`

                <aspect name="kyp:siteContent">
                        <title>Site du contenu</title>
                        <properties>
                                <property name="kyp:siteShortName">
                                   <type>d:text</type>
                                </property>
                        </properties>
                </aspect>
`
**_alfresco-global.properties :_**

`
dir.root=/opt/alfresco-5.0.d/alf_data

[...]

simpleContentStores.enabled=true
simpleContentStores.rootStore=siteRoutingStore
simpleContentStores.customStores=siteRoutingStore,siteAB1Store,siteAB2Store,siteAB3Store

simpleContentStores.customStore.siteRoutingStore.type=selectorPropertyStore
simpleContentStores.customStore.siteRoutingStore.value.selectorClassName=kyp:siteContent
simpleContentStores.customStore.siteRoutingStore.value.selectorPropertyName=kyp:siteShortName

simpleContentStores.customStore.siteRoutingStore.ref.fallbackStore=fileContentStore
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB1=siteAB1Store
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB2=siteAB2Store
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB3=siteAB3Store

simpleContentStores.customStore.siteRoutingStore.value.moveStoresOnChange=true

simpleContentStores.customStore.siteAB1Store.type=standardFileStore
simpleContentStores.customStore.siteAB1Store.value.protocol=site-ab1-store
simpleContentStores.customStore.siteAB1Store.value.rootDirectory=${dir.root}/site-ab1-contentstore

simpleContentStores.customStore.siteAB2Store.type=standardFileStore
simpleContentStores.customStore.siteAB2Store.value.protocol=site-ab2-store
simpleContentStores.customStore.siteAB2Store.value.rootDirectory=${dir.root}/site-ab2-contentstore

simpleContentStores.customStore.siteAB3Store.type=standardFileStore
simpleContentStores.customStore.siteAB3Store.value.protocol=site-ab3-store
simpleContentStores.customStore.siteAB3Store.value.rootDirectory=${dir.root}/site-ab3-contentstore

`


**_content store configuration :_**
ll /opt/alfresco-5.0.d/alf_data
drwxr-xr-x  3 root     root 4096 août  31 10:16 contentstore/
drwxr-xr-x  3 root     root 4096 sept. 17 04:00 contentstore.deleted/
drwxr-xr-x  2 root     root 4096 août  31 10:14 keystore/
drwx------  3 root     root 4096 sept.  7 09:37 oouser/
drwx------ 15 postgres root 4096 sept. 20 00:27 postgresql/
drwxr-xr-x  2 root     root 4096 sept. 19 00:03 site-ab1-contentstore/
drwxr-xr-x  2 root     root 4096 sept. 19 00:03 site-ab2-contentstore/
drwxr-xr-x  2 root     root 4096 sept. 19 00:03 site-ab3-contentstore/
drwxr-xr-x  5 root     root 4096 août  31 10:15 solr4/
drwxr-xr-x  4 root     root 4096 sept.  8 04:00 solr4Backup/

Then, I defined a rule on the specific folder for which I want the content to be stored in the store AB1 :

![image](https://cloud.githubusercontent.com/assets/21683549/18732440/7e616e86-8063-11e6-8622-2af37b996031.png)



However, I will test your new version of the simple-content-stores add-on.

Thanks again for your help.

Arnaud.

**Comment by ArnoLP** [comment]
Hello Axel,

In fact I discovered that the package was not really installed.

So I installed it correctly, and now I get errors when launching Alfresco (I get the error when I set the parameter simpleContentStores.enabled=true
If I set it to false, Alfresco starts correctly.

My version of Alfresco is 5.0.d Community. I just installed the package on the "official" version, without any modification.
OS is Ubuntu 14.04 LTS.

I copy here some part of the alfresco.log. if you need more info, please tell me.

2016-09-22 23:53:24,248 ERROR [org.springframework.web.context.ContextLoader] [localhost-startStop-1] Context initialization failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'authorityBridgeTableCache' defined in class path resource [alfresco/cache-context.xml]: Cannot resolve reference to bean 'tenantAdminService' while setting bean property 'tenantAdminService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'tenantAdminService' defined in URL [jar:file:/opt/alfresco-5.0.d/tomcat/webapps/alfresco/WEB-INF/lib/alfresco-repository-5.0.d.jar!/alfresco/mt/mt-admin-context.xml]: Cannot resolve reference to bean 'dbNodeServiceImpl' while setting bean property 'nodeService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'dbNodeService' defined in class path resource [alfresco/node-services-context.xml]: Cannot resolve reference to bean 'permissionServiceImpl' while setting bean property 'permissionService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'permissionServiceImpl' defined in class path resource [alfresco/public-services-security-context.xml]: Cannot resolve reference to bean 'authorityService' while setting bean property 'authorityService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'authorityService' defined in class path resource [alfresco/authority-services-context.xml]: Cannot resolve reference to bean 'personService' while setting bean property 'personService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'personService' defined in class path resource [alfresco/authentication-services-context.xml]: Cannot resolve reference to bean 'personServicePermissionsManager' while setting bean property 'permissionsManager'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'personServicePermissionsManager' defined in class path resource [alfresco/authentication-services-context.xml]: Cannot resolve reference to bean 'ownableService' while setting bean property 'ownableService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'ownableService' defined in class path resource [alfresco/ownable-services-context.xml]: Cannot resolve reference to bean 'renditionService' while setting bean property 'renditionService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'renditionService' defined in class path resource [alfresco/rendition-services-context.xml]: Cannot resolve reference to bean 'ActionService' while setting bean property 'actionService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'ActionService' defined in class path resource [alfresco/public-services-context.xml]: Cannot resolve reference to bean 'actionService' while setting bean property 'target'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'actionService' defined in class path resource [alfresco/action-services-context.xml]: Cannot resolve reference to bean 'NodeService' while setting bean property 'nodeService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'NodeService': FactoryBean threw exception on object creation; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'AuditMethodInterceptor' defined in class path resource [alfresco/public-services-context.xml]: Cannot resolve reference to bean 'auditComponent' while setting bean property 'auditComponent'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'auditComponent' defined in class path resource [alfresco/audit-services-context.xml]: Cannot resolve reference to bean 'auditDAO' while setting bean property 'auditDAO'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'auditDAO' defined in class path resource [alfresco/dao/dao-context.xml]: Cannot resolve reference to bean 'contentService' while setting bean property 'contentService'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in class path resource [alfresco/content-services-context.xml]: Cannot resolve reference to bean 'siteRoutingStore' while setting bean property 'store'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'siteRoutingStore': Invocation of init method failed; nested exception is org.alfresco.error.AlfrescoRuntimeException: 08220000 Property 'classQName' has not been set: de.axelfaust.alfresco.simplecontentstores.repo.store.SelectorPropertyContentStore@6ca59947 (class de.axelfaust.alfresco.simplecontentstores.repo.store.SelectorPropertyContentStore)
 
Thanks for your help.

Kind Regards,

Arnaud.

**Comment by AFaust** [comment]
You have not set the property "classQName" on your property selector store instance "siteRoutingStore". The classQName is required when you set "moveStoresOnChange" to true. It is used for registering the behaviour / policy and should be the name of the aspect that defines your custom property.

**Comment by ArnoLP** [comment]
Hello Axel, 

Thanks for your feedback.

1) I have re-checked what I did to create my model, and it seems to be OK.
If I set simpleContentStores.enabled=false, then Alfresco works fine, and I can define that a content has the defined aspect, and with the defined property value.
So it should mean that the model has been well activated in Alfresco. 

If I set simpleContentStores.enabled=true, then I get the error.

So I don't understand what I missed.
By the way, I got from you another message :
"
You have not set the property "classQName" on your property selector store instance "siteRoutingStore". The classQName is required when you set "moveStoresOnChange" to true. It is used for registering the behaviour / policy and should be the name of the aspect that defines your custom property.
"
I don't understand what I have to do with it, and if as explained, I shoudl add a new line in the alfresco-global.properties defining the "classQName" property.
Can you re-explain, please ?

2) Then, I tried to use your last version of the add-on.

If I create a site before the add-on is activated, and then activate the add-on defining this site as a custom store, the documents I upload in this site are well stored in the specific content store. It's nice !
But then, I am unable to create any other site, or to upload documents to existing "standard" sites.

Is it still somewhere linked to my configuration, or do you already know this issue ?

I used a standard Alfresco 5.0.d community version without any specific parameter neither development.

If it can help, here is an extract of the alfresco.log :

2016-09-27 12:17:44,336 ERROR [org.alfresco.repo.web.scripts.RepositoryContainer] [http-apr-8080-exec-7] Server error (08270041)
org.alfresco.error.AlfrescoRuntimeException: 08270041 08270040 Error during executing call on ContentStore API
...
Caused by: org.alfresco.service.cmr.repository.ContentIOException: 08270040 Error during executing call on ContentStore API
        at de.axelfaust.alfresco.simplecontentstores.repo.aop.InitContentStoreContextInterceptor$1.execute(InitContentStoreContextInterceptor.java:87)
        at de.axelfaust.alfresco.simplecontentstores.repo.store.context.ContentStoreContext.executeInNewContext(ContentStoreContext.java:129)
        at de.axelfaust.alfresco.simplecontentstores.repo.aop.InitContentStoreContextInterceptor.invoke(InitContentStoreContextInterceptor.java:57)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at de.axelfaust.alfresco.simplecontentstores.repo.aop.ContentStoreCapsEmulatingInterceptor.invoke(ContentStoreCapsEmulatingInterceptor.java:18
2)
...
Caused by: org.alfresco.service.cmr.repository.ContentIOException: 08270039 Failed to get writer for URL: null
        at de.axelfaust.alfresco.simplecontentstores.repo.store.FileContentStore.getWriterInternal(FileContentStore.java:486)
        at de.axelfaust.alfresco.simplecontentstores.repo.store.SiteAwareMultiDirectoryFileContentStore.getWriterInternal(SiteAwareMultiDirectoryFileContentStore.java:236)
...
Caused by: java.lang.NullPointerException
        at de.axelfaust.alfresco.simplecontentstores.repo.store.SiteAwareMultiDirectoryFileContentStore.determineEffectiveContentUrl(SiteAwareMultiDirectoryFileContentStore.java:406)
        at de.axelfaust.alfresco.simplecontentstores.repo.store.SiteAwareMultiDirectoryFileContentStore.createNewFileStoreUrl(SiteAwareMultiDirectoryFileContentStore.java:248)
        at de.axelfaust.alfresco.simplecontentstores.repo.store.FileContentStore.createNewFile(FileContentStore.java:499)
        at de.axelfaust.alfresco.simplecontentstores.repo.store.FileContentStore.getWriterInternal(FileContentStore.java:453)
        ... 78 more

Thanks for your help.

Kind Regards,

Arnaud.

**Comment by AFaust** [comment]
1) The other message was a just previous version of the same comment and can be ignored.
If you are sure that the model is really properly activated, ok, but I have just tested the entire configuration with a aspect/property from the default model and everything works without a problem.

2) It appears you have not tested with the latest state. The error you are getting was corrected on Sep 24th (https://github.com/AFaust/simple-content-stores/commit/20da7d9f7be0768683ddb5a1447eb492d6ab8f23#diff-716ca8e363f15448da2ae23399ab0e10L406).

**Comment by ArnoLP** [comment]
Thanks a lot, both cases work now ! Real nice work.
For the 1st case, I used as you explained, the aspect you put in your example (from the default model), and it works fine.

I still have a question :
If I want to combine both cases, it currently doesn't work. But I suppose my configuration in alfresco-global.properties is not correct. Should I use one the new content store types as the root store (site-aware routing content store or any other one) ? It's not clear in my mind.

Could you help, please ?

Thanks in advance.

Arnaud.

here is what I configured.

simpleContentStores.enabled=true

simpleContentStores.rootStore=siteFileContentStore
simpleContentStores.customStores=siteFileContentStore, siteRoutingStore,siteAB2Store,siteAB3Store

simpleContentStores.customStore.siteFileContentStore.type=siteAwareMultiDirectoryFileStore
# default root directory for all other content
simpleContentStores.customStore.siteFileContentStore.value.rootDirectory=${dir.root}/contentstore

# site specific root directory + protocol
simpleContentStores.customStore.siteFileContentStore.map.rootAbsolutePathsBySite.value.essai=${dir.root}/site-ab1-contentstore
simpleContentStores.customStore.siteFileContentStore.map.protocolsBySite.value.essai=my-site-store

simpleContentStores.customStore.siteFileContentStore.map.rootAbsolutePathsBySite.value.sitealp=${dir.root}/sitealp-contentstore
simpleContentStores.customStore.siteFileContentStore.map.protocolsBySite.value.sitealp=sitealp-store

simpleContentStores.customStore.siteRoutingStore.type=selectorPropertyStore
simpleContentStores.customStore.siteRoutingStore.value.selectorClassName=cm:storeSelector
simpleContentStores.customStore.siteRoutingStore.value.selectorPropertyName=cm:storeName

simpleContentStores.customStore.siteRoutingStore.ref.fallbackStore=fileContentStore
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB2=siteAB2Store
simpleContentStores.customStore.siteRoutingStore.map.storeBySelectorPropertyValue.ref.siteAB3=siteAB3Store

simpleContentStores.customStore.siteRoutingStore.value.moveStoresOnChange=true

simpleContentStores.customStore.siteAB2Store.type=standardFileStore
simpleContentStores.customStore.siteAB2Store.value.protocol=site-ab2-store
simpleContentStores.customStore.siteAB2Store.value.rootDirectory=${dir.root}/site-ab2-contentstore

simpleContentStores.customStore.siteAB3Store.type=standardFileStore
simpleContentStores.customStore.siteAB3Store.value.protocol=site-ab3-store
simpleContentStores.customStore.siteAB3Store.value.rootDirectory=${dir.root}/site-ab3-contentstore

**Comment by AFaust** [comment]
Yes, the root store needs to be a different store if you want to combine both concepts. Basically you need to configure a routing store as the root store - a fileContentStore (site-aware or not) then should be used for the fallback store as well as the storeBySelectorPropertyValue.

I have used the following configuration to test the current state for my response from yesterday. This creates a selectorPropertyStore as the root, which routes to a siteRoutingStore which in turn routes to both a standardFileStore and a siteAwareFileStore. It is a really contrived example but shows how all the stores can be used to define any number of routing layers...

```
simpleContentStores.enabled=true
simpleContentStores.rootStore=selectorPropertyStore
simpleContentStores.customStores=selectorPropertyStore,siteFileContentStore,siteRoutingStore,thirdSiteStore

simpleContentStores.customStore.selectorPropertyStore.type=selectorPropertyStore
simpleContentStores.customStore.selectorPropertyStore.map.storeBySelectorPropertyValue.ref.dummy=siteRoutingStore
simpleContentStores.customStore.selectorPropertyStore.ref.fallbackStore=siteRoutingStore
simpleContentStores.customStore.selectorPropertyStore.value.selectorClassName=cm:emailed
simpleContentStores.customStore.selectorPropertyStore.value.selectorPropertyName=cm:originator
simpleContentStores.customStore.selectorPropertyStore.value.moveStoresOnChange=true

simpleContentStores.customStore.siteRoutingStore.type=siteRoutingStore
simpleContentStores.customStore.siteRoutingStore.map.storeBySite.ref.third-site=thirdSiteStore
simpleContentStores.customStore.siteRoutingStore.ref.fallbackStore=siteFileContentStore
simpleContentStores.customStore.siteRoutingStore.value.moveStoresOnNodeMoveOrCopy=true

simpleContentStores.customStore.thirdSiteStore.type=standardFileStore
simpleContentStores.customStore.thirdSiteStore.value.rootDirectory=${dir.root}/third-site-store
simpleContentStores.customStore.thirdSiteStore.value.protocol=third-site-store

simpleContentStores.customStore.siteFileContentStore.type=siteAwareMultiDirectoryFileStore
simpleContentStores.customStore.siteFileContentStore.value.rootDirectory=${dir.contentstore}
simpleContentStores.customStore.siteFileContentStore.map.rootAbsolutePathsBySite.value.my-site=${dir.root}/my-site-contentstore
simpleContentStores.customStore.siteFileContentStore.map.protocolsBySite.value.my-site=my-site-store
simpleContentStores.customStore.siteFileContentStore.value.moveStoresOnNodeMoveOrCopy=true
simpleContentStores.customStore.siteFileContentStore.value.useSiteFolderInGenericDirectories=true
```

**Comment by ArnoLP** [comment]
Thanks for your feedback, Axel.
(I still have to test this kind of configuration).

I have another question about this add-on : I saw that there is a compressing content store. Do you have in the planned content stores, a "crypted" content store, which would crypt (and transparently decrypt) content ? Using truecrypt as an example ?

Thanks again for this very valuable add-on.

**Comment by AFaust** [comment]
A simple encrypted content store is generally planned for. But I don't think it would use TrueCrypt (a discontinued library / project).

**Comment by ArnoLP** [comment]
Do you have a rough idea when you plan to work on this encrypted content store ? Or not really planned yet ?
Thanks again.

**Comment by AFaust** [comment]
Since it recently came up in the discussion with a prospective customer I have taken the time and implemented an encrypted content store in the last two days.

**Comment by AFaust** [comment]
Original issue was successfully addressed quite a while ago - closing as long as there is no further input.
